Blog for vllm optimizations on Intel GPU

rogerxfeng8 opened the pull request:
The vllm optimizations and support on Intel GPU, using ARC Pro B series cards to showcase the performance.

rogerxfeng8 committed:
blog for vllm optimizations on Intel GPU/XPU

Update the vllm optimizations and support on Intel GPU, using ARC Pro B series product.

Signed-off-by: roger feng <roger.feng@intel.com>

hmellor commented on _posts/2025-11-11-intel-arc-pro-b.md:
File name after the date  becomes the URL. It would be better to simplify to something like `*-31-intel-arc-pro-b.md` for a prettier URL

hmellor commented on _posts/2025-11-11-intel-arc-pro-b.md:
```diff
@@ -0,0 +1,128 @@
+
+
+[Intel® Arc™ Pro B-Series GPU Family](https://www.intel.com/content/www/us/en/products/docs/discrete-gpus/arc/workstations/b-series/overview.html) GPUs deliver powerful AI capabilities with a focus on accessibility and exceptional price-to-performance ratios. Their large memory capacity and scalability with multi-GPU setups make it possible to run the latest, large and capable AI models locally,  ...
+
+vLLM is at the core of the software stack enabling fast and cost-effective LLM serving on Intel Arc Pro B-Series GPUs. Over the past few months, Intel developers have been actively collaborating with the vLLM community to enable and optimize key features and ensure seamless performance with multi-GPU scaling and PCIe P2P data transfer on Intel Arc Pro B-Series GPUs.
```
Duplicated paragraph
```suggestion
```

rogerxfeng8 replied:
nice catch

hmellor commented on _posts/2025-10-31-vllm-on-intel-Arc-pro-B.md:
```diff
@@ -0,0 +1,128 @@
+author: "Intel vLLM Team"
+---
+
+
```
```suggestion

```

hmellor commented on _posts/2025-10-31-vllm-on-intel-Arc-pro-B.md:
```diff
@@ -0,0 +1,128 @@
+
+vLLM is at the core of the software stack enabling fast and cost-effective LLM serving on Intel Arc Pro B-Series GPUs. Over the past few months, Intel developers have been actively collaborating with the vLLM community to enable and optimize key features and ensure seamless performance with multi-GPU scaling and PCIe P2P data transfer on Intel Arc Pro B-Series GPUs.
+
+Based on vLLM v1 engine, Intel® Arc™ Pro B-series GPUs provide vLLM key features and optimizations including:
```
V0 no longer exists, so V1 is no longer a selling point

```suggestion
Intel® Arc™ Pro B-series GPUs provide vLLM key features and optimizations including:

```

hmellor commented on _posts/2025-10-31-vllm-on-intel-Arc-pro-B.md:
```diff
@@ -0,0 +1,128 @@
+vLLM is at the core of the software stack enabling fast and cost-effective LLM serving on Intel Arc Pro B-Series GPUs. Over the past few months, Intel developers have been actively collaborating with the vLLM community to enable and optimize key features and ensure seamless performance with multi-GPU scaling and PCIe P2P data transfer on Intel Arc Pro B-Series GPUs.
+
+Based on vLLM v1 engine, Intel® Arc™ Pro B-series GPUs provide vLLM key features and optimizations including:
+-	Solid inference performance for DeepSeek distilled llama/qwen models
```
```suggestion
-	Solid inference performance for DeepSeek distilled Llama/Qwen models
```

hmellor commented on _posts/2025-10-31-vllm-on-intel-Arc-pro-B.md:
```diff
@@ -0,0 +1,128 @@
+-	Solid inference performance for DeepSeek distilled llama/qwen models
+-	Long context length (>50K) with good scaling on batch size
+-	Support for embedding, reranker, pooling models
+-	Support multi-modality models
```
```suggestion
-	Support for multi-modal models
```

hmellor commented on _posts/2025-10-31-vllm-on-intel-Arc-pro-B.md:
```diff
@@ -0,0 +1,128 @@
+-	Support for embedding, reranker, pooling models
+-	Support multi-modality models
+-	Well optimized Mixture of Experts (MoE) models (GPT-OSS, DeepSeek-v2-lite, Qwen3-30B-A3B etc)
+-	By-layer online quantization to reduce the required GPU memory
```
```suggestion
-	Per-layer online quantization to reduce the required GPU memory
```

hmellor commented on _posts/2025-10-31-vllm-on-intel-Arc-pro-B.md:
```diff
@@ -0,0 +1,128 @@
+-	Tool calling
+-	Mixed precision support for BF16, FP16, INT4 and FP8 vLLM recipes
+
+
+## Advanced Optimizations for MoE Models
```
```suggestion

## Advanced Optimizations for MoE Models

```

hmellor commented on _posts/2025-10-31-vllm-on-intel-Arc-pro-B.md:
```diff
@@ -0,0 +1,128 @@
+However, naive implementations of MoE GEMM operations can suffer from significant efficiency bottlenecks. The typical approach, where individual GEMM kernels are sequentially launched per iteration on a for-loop, generates excessive kernel launch overhead and introduces substantial scheduling latency. Furthermore, since expert routing decisions are produced by the gating network, GEMM operations  ...
+
+Targeting these limitations of MoE GEMM we designed a persistent zero gap kernel which achieved over 80% efficiency of hardware capacity of Intel® Arc™ Pro B60 GPU.
+## Optimization 1. Single kernel launched in persistent loop
```
```suggestion

### Optimization 1. Single kernel launched in persistent loop

```

hmellor commented on _posts/2025-10-31-vllm-on-intel-Arc-pro-B.md:
```diff
@@ -0,0 +1,128 @@
+
+Intel® Arc™ Pro B60 GPU has 20 XeCores, each with identical resources that can host multiple SYCL groups. In our design, we launch two groups per XeCore to balance compute and memory bandwidth needs.
+
+## Optimization 2. Dynamic balancing of computing groups
```
```suggestion
### Optimization 2. Dynamic balancing of computing groups

```

hmellor commented on _posts/2025-10-31-vllm-on-intel-Arc-pro-B.md:
```diff
@@ -0,0 +1,128 @@
+
+We mitigate the effect by letting each group compete for the next job through an atomic number. Whoever finishes computing one GEMM block will get a rank from the atomic number who decides which next block it’ll take. In this case, we eliminated small gaps in kernel looping and achieved perfect scheduling among all scenarios of experts routing.
+
+## Optimization 3. Fast MXFP4 to BFLOAT16 algorithm with prepack for memory load efficiency
```
```suggestion
### Optimization 3. Fast MXFP4 to BFLOAT16 algorithm with prepack for memory load efficiency

```

hmellor commented on _posts/2025-10-31-intel-arc-pro-b.md:
```diff
@@ -0,0 +1,128 @@
+For a concrete example, there are 40 groups to crunch 200 GEMM blocks, static stride will result that group 0 loop through 0, 40, 80, ... group 1 loop through 1, 41, 81, etc. A caveat is that due to the nature of MoE, each GEMM block may not have same amount of compute intensity. Also, randomized access patterns will let certain groups finish work faster than others. This will limit efficiency in ...
+
+![thread load](/assets/figures/2025-vllm-on-intel-arc/thread-load1.png)
+![thread load](/assets/figures/2025-vllm-on-intel-arc/thread-load2.png)
```
There is no separation between these images in the rendered blog

rogerxfeng8 replied:
Would like to put the two figures side by side for a straight-forward comparison. Is this the right markdown format?

hmellor replied:
For side by side you can either put them in a table or use raw html.

I'm not sure what the table option would look like in the blog but the md would look something like:

```md
| Before | After |
|---|---|
| ![thread load](/assets/figures/2025-vllm-on-intel-arc/thread-load1.png) | ![thread load](/assets/figures/2025-vllm-on-intel-arc/thread-load2.png) |

hmellor commented on _posts/2025-10-31-vllm-on-intel-Arc-pro-B.md:
```diff
@@ -0,0 +1,128 @@
+
+## Optimization 3. Fast MXFP4 to BFLOAT16 algorithm with prepack for memory load efficiency
+Prepacking has long been known to improve memory load efficiency. For 4-bit memory loads, a hardware-friendly format can increase efficiency by up to 30%, as observed in our case. Also, naive FP4 to BF16 incurs too many instructions which prompt a need for better alternative (borrow from oneDNN, stride E2M1 encoding on single precision E/M position and multiple the scale difference between two ty ...
+Bitcast-bf16 ((x << 12) >> 6 & 0x81c0) * 2^126
```
```suggestion

`Bitcast-bf16 ((x << 12) >> 6 & 0x81c0) * 2^126`

```

hmellor replied:
Should we code format this line?

hmellor commented on _posts/2025-11-11-intel-arc-pro-b.md:
```diff
@@ -0,0 +1,128 @@
+With 24GB of high-bandwidth VRAM, 456 GB/s memory bandwidth and 160 Intel® Xe Matrix Extensions (Intel® XMX) AI engines, Intel Arc Pro B-Series GPUs offers good hardware capacity for the optimization of high touch models on vLLM . The full support model list can be found at [intel/ai-containers](https://github.com/intel/ai-containers/blob/main/vllm/0.10.2-xpu.md#supported-models)
+
+DeepSeek distilled models sized from 8B to 70B are optimized for good output token throughput on a system with eight Intel® Arc™ Pro GPUs.
+![model perf](/assets/figures/2025-vllm-on-intel-arc/perf-figure1.png)  
```
```suggestion

![model perf](/assets/figures/2025-vllm-on-intel-arc/perf-figure1.png)  
```

hmellor commented on _posts/2025-11-11-intel-arc-pro-b.md:
```diff
@@ -0,0 +1,128 @@
+Figure 1: FP8 model output token throughput with max concurrency under SLA on a system configured with 8 Intel® Arc™ Pro B60 GPU cards
+
+The system sustains less than 100 ms next token latencies with good concurrency load.  
+![model perf](/assets/figures/2025-vllm-on-intel-arc/perf-figure2.png)  
```
```suggestion

![model perf](/assets/figures/2025-vllm-on-intel-arc/perf-figure2.png)  
```

hmellor commented on _posts/2025-10-31-vllm-on-intel-Arc-pro-B.md:
```diff
@@ -0,0 +1,128 @@
+Figure 2: Qwen-32B next token latency with increasing number of prompts on a system configured with 4 Intel® Arc™ Pro B60 GPU cards
+
+The model inference maintains consistent next-token latency across a wide range of input sequence lengths, scaling from 1K to over 40K tokens. This performance is underpinned by highly optimized flash attention kernels that parallelize operations across the sequence length dimension.  
+![model perf](/assets/figures/2025-vllm-on-intel-arc/perf-figure3.png)  
```
```suggestion

![model perf](/assets/figures/2025-vllm-on-intel-arc/perf-figure3.png)  
```

hmellor commented on _posts/2025-10-31-vllm-on-intel-Arc-pro-B.md:
```diff
@@ -0,0 +1,128 @@
+Figure 2: Qwen-32B next token latency with increasing number of prompts on a system configured with 4 Intel® Arc™ Pro B60 GPU cards
+
+The model inference maintains consistent next-token latency across a wide range of input sequence lengths, scaling from 1K to over 40K tokens. This performance is underpinned by highly optimized flash attention kernels that parallelize operations across the sequence length dimension.  
+![model perf](/assets/figures/2025-vllm-on-intel-arc/perf-figure3.png)  
+Figure 3: TTFT/TPOT for llama-70B single batch with long context input from 1K to 40K sequences on a system configured with 8 Intel® Arc™ Pro B60 GPU cards
+
+
```
```suggestion

```

hmellor commented on _posts/2025-10-31-vllm-on-intel-Arc-pro-B.md:
```diff
@@ -0,0 +1,128 @@
+| GPT-OSS-120b	|MXFP4	|4	|2048/2048	|50	|8.11	|41.98	|1085.58|
+| GPT-OSS-120b	|MXFP4	|4	|5120/5120	|20	|8.60	|30.60	|619.10 |
+
+Table 1: GPT-OSS vLLM inference throughput using 1-4 GPUs on x8 Intel® Arc™ Pro B-series System
```
```suggestion
Table 1: GPT-OSS vLLM inference throughput using 1-4 GPUs on x8 Intel® Arc™ Pro B-series System.
```

hmellor commented on _posts/2025-10-31-vllm-on-intel-Arc-pro-B.md:
```diff
@@ -0,0 +1,128 @@
+`$ vllm serve openai/gpt-oss-120b --dtype=bfloat16 --enforce-eager --port 8000 --host 0.0.0.0 --trust-remote-code --gpu-memory-util=0.9 --no-enable-prefix-caching --max-num-batched-tokens=8192 --disable-log-requests --max-model-len=16384 --block-size 64 -tp 4`
+
+\# Start another shell and run the benchmarking  
+`$ vllm bench serve --model --model openai/gpt-oss-120b --dataset-name sonnet --dataset-path="./benchmarks/sonnet.txt" --sonnet-input-len=1024 --sonnet-output-len=1024 --ignore-eos --num-prompt 1 --trust_remote_code --request-rate inf --backend vllm --port=8000 --host 0.0.0.0`
```
````suggestion
Get the released docker image with command  

```bash
docker pull intel/vllm:0.10.2-xpu
```

Instantiate a docker container with command  

```bash
docker run -t -d --shm-size 10g --net=host --ipc=host --privileged -v /dev/dri/by-path:/dev/dri/by-path --name=vllm-test --device /dev/dri:/dev/dri --entrypoint= intel/vllm:0.10.2-xpu /bin/bash
```

Run the vllm server with gpt-oss-120b on 4 Intel® Arc™ Pro B60 cards  

```bash
vllm serve openai/gpt-oss-120b --dtype=bfloat16 --enforce-eager --port 8000 --host 0.0.0.0 --trust-remote-code --gpu-memory-util=0.9 --no-enable-prefix-caching --max-num-batched-tokens=8192 --disable-log-requests --max-model-len=16384 --block-size 64 -tp 4
```

Start another shell and run the benchmarking  

```bash
vllm bench serve --model --model openai/gpt-oss-120b --dataset-name sonnet --dataset-path="./benchmarks/sonnet.txt" --sonnet-input-len=1024 --sonnet-output-len=1024 --ignore-eos --num-prompt 1 --trust_remote_code --request-rate inf --backend vllm --port=8000 --host 0.0.0.0
```
````

simon-mo commented:
@codex review

rogerxfeng8 committed:
format and phrase update

address the review comments

Signed-off-by: roger feng <roger.feng@intel.com>

rogerxfeng8 commented:
@hmellor Thanks for the review! Comments are addressed.

jikunshang commented:
@hmellor @simon-mo can you take a review again?

hmellor commented:
Looks nice, just 2 unresolved comments about the side by side images and the formatting of a line which might look nicer in code format.

rogerxfeng8 committed:
Format update

Minor format changes on the side-by-side figures and using code format for formula

Signed-off-by: roger feng <roger.feng@intel.com>

rogerxfeng8 commented:
> Looks nice, just 2 unresolved comments about the side by side images and the formatting of a line which might look nicer in code format.

The two comments are addressed. Looks better. Thanks!

hmellor reviewed (approved):
LGTM!

Now you can just update the file name to today's date and we can publish!

rogerxfeng8 committed:
update file name

Signed-off-by: roger feng <roger.feng@intel.com>

hmellor committed:
Merge branch 'main' into dev
